pkg/apis/channels/v1alpha1/subscription_types.go

```diff
@@ -57,14 +64,53 @@ type SubscriptionSpec struct {
 	// +optional
 	Generation int64 `json:"generation,omitempty"`
 
-	// Channel is the name of the channel to subscribe to.
-	Channel string `json:"channel"`
-
-	// Subscriber is the name of the subscriber service DNS name.
-	Subscriber string `json:"subscriber"`
-
-	// Target service DNS name for replies returned by the subscriber.
-	ReplyTo string `json:"replyTo,omitempty"`
+	// Reference to an object that will be used to create the subscription
+	// for receiving events. The object must have spec.subscriptions
+	// list which will then be modified accordingly.
+	//
+	// This object must fulfill the Subscribable contract.
+	//
+	// You can specify only the following fields of the ObjectReference:
+	//   - Kind
+	//   - APIVersion
+	//   - Name
+	From *corev1.ObjectReference `json:"from,omitempty"`
+
+	// Processor is reference to (optional) function for processing events.
+	// Events from the From channel will be delivered here and replies
+	// are sent to To channel.
+	//
+	// This object must fulfill the Targetable contract.
+	//
+	// Reference to an object that will be used to deliver events for
+	// (optional) processing before sending them to To for further
+	// if specified for additional Subscriptions to then subscribe
+	// to these events for further processing.
+	//
+	// For example, this could be a reference to a Route resource
+	// or a Service resource.
+	// TODO: Specify the required fields the target object must
+	// have in the status.
+	// You can specify only the following fields of the ObjectReference:
+	//   - Kind
+	//   - APIVersion
+	//   - Name
+	// +optional
+	Processor *corev1.ObjectReference `json:"processor,omitempty"`
```

Comment:
Is the idea that you would need to create a k8s Service to be able to call something which is not on the k8s cluster (e.g. a VM)? Previously we had the idea that a targetable was either a DNS name or an ObjectReference.

Reply:
Yeah, so, this is a bit interesting in a sense, that I'd expect that if you're hitting something that's outside the cluster, just a DNS entry is probably not going to cut it. I'd expect things like credentials to be involved on that, so as not to just be wide open to everything. I could hoist this into a higher level object to give us some future proofing for now.

Reply:
Another option is to allow the reference for a targetable to also include a k8s Service. It doesn't implement the targetable interface, but we know how to shoehorn it into what we need.

Reply:
@scothis when we talked about this wrt Targetable, thinking was that a service was not enough, that it would actually have to be a URI. So, I'm hoisting that up.